<a href="https://colab.research.google.com/github/MochamadPhilliaWibowo/Machine-Learning-Class/blob/main/Tugas%20Week%2013/BAB%202/Tokenizers_(PyTorch).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Nama : Mochamad Phillia Wibowo

NIM : 1103204191

Kelas : Machine Learning (TK-44-G04)

Tugas 13 NLP HUGGING FACE (BAB 2)

# Tokenizers (PyTorch)

Install the Transformers, Datasets, and Evaluate libraries to run this notebook.

In [1]:
!pip install datasets evaluate transformers[sentencepiece]

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 542.1/542.1 kB 4.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 5.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 6.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.9/64.9 kB 5.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 6.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 7.0 MB/s eta 0:00:00
  Attempting uninstall: requests
    Found existing installation: requests 2.31.0
    Uninstalling requests-2.31.0:
      Successfully uninstalled requests-2.31.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires requests==2.31.0, but you have requests 2.32.3 which is incompatible.


Code di atas adalah perintah untuk menginstal beberapa pustaka Python menggunakan pip, manajer paket standar untuk Python. Perintah ini melakukan instalasi tiga pustaka:

datasets: Ini adalah pustaka yang menyediakan akses mudah ke berbagai kumpulan data untuk tugas-tugas pembelajaran mesin dan pemrosesan bahasa alami.

evaluate: Ini mungkin adalah pustaka khusus yang digunakan untuk mengevaluasi model-model pembelajaran mesin atau pemrosesan bahasa alami, tetapi tanpa informasi lebih lanjut, sulit untuk mengkonfirmasi.

transformers[sentencepiece]: Ini adalah pustaka yang populer digunakan untuk pemrosesan bahasa alami, terutama dalam bidang pemodelan bahasa dan transfer pembelajaran. Opsi [sentencepiece] menunjukkan bahwa instalasi juga akan mencakup dukungan untuk SentencePiece, sebuah pustaka tokenisasi teks multibahasa.

Dengan menjalankan perintah ini, pustaka-pustaka ini akan diinstal dan siap digunakan dalam pengembangan aplikasi Python Anda.

In [2]:
tokenized_text = "Jim Henson was a puppeteer".split()
print(tokenized_text)

['Jim', 'Henson', 'was', 'a', 'puppeteer']


Baris ini menginisialisasi variabel tokenized_text yang berisi token-token dari kalimat "Jim Henson was a puppeteer" yang telah dipisahkan menggunakan spasi.

- "Jim Henson was a puppeteer".split(): Baris ini menggunakan metode split() untuk memisahkan kalimat menjadi token-token dengan menggunakan spasi sebagai pemisah. Hasilnya adalah sebuah daftar yang berisi token-token individu dari kalimat tersebut.

In [3]:
from transformers import BertTokenizer

tokenizer = BertTokenizer.from_pretrained("bert-base-cased")

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/436k [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

Baris pertama mengimpor kelas BertTokenizer dari pustaka Transformers. BertTokenizer adalah kelas yang digunakan untuk tokenisasi teks menggunakan model BERT.

Baris kedua menginisialisasi sebuah tokenizer dengan menggunakan metode from_pretrained() dari kelas BertTokenizer. Metode ini memuat tokenizer yang sesuai dengan model yang telah ditentukan dalam argumen, dalam hal ini "bert-base-cased". Ini akan memuat tokenizer yang menggunakan kamus token yang cocok dengan model BERT berbahasa Inggris yang berbasis huruf besar (cased).

Dengan demikian, setelah kode dieksekusi, variabel tokenizer akan berisi tokenizer yang siap digunakan untuk memproses teks menggunakan model BERT yang telah di-pretrained.

In [4]:
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")

Baris pertama mengimpor kelas AutoTokenizer dari pustaka Transformers. AutoTokenizer adalah kelas yang digunakan untuk menginisialisasi dan memuat tokenizer yang sesuai dengan model yang akan digunakan.

Baris kedua menginisialisasi sebuah tokenizer dengan menggunakan metode from_pretrained() dari kelas AutoTokenizer. Metode ini memuat tokenizer yang sesuai dengan model yang ditentukan dalam argumen, dalam hal ini "bert-base-cased". Ini akan memuat tokenizer yang sesuai dengan model BERT berbahasa Inggris yang berbasis huruf besar (cased).

Dengan menggunakan AutoTokenizer, kita memungkinkan untuk mengganti model tanpa harus secara eksplisit menentukan tipe tokenizer, karena AutoTokenizer akan secara otomatis memilih tokenizer yang sesuai berdasarkan model yang diberikan dalam argumen from_pretrained().

In [5]:
tokenizer("Using a Transformer network is simple")

{'input_ids': [101, 7993, 170, 13809, 23763, 2443, 1110, 3014, 102], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1]}

Kode yang Anda berikan menggunakan tokenizer yang telah diinisialisasi sebelumnya untuk memproses teks "Using a Transformer network is simple". Mari kita bahas lebih detail:

tokenizer("Using a Transformer network is simple"): Baris ini menggunakan tokenizer yang telah diinisialisasi sebelumnya untuk memproses teks yang diberikan sebagai argumen. Tokenizer akan menghasilkan output berupa representasi token dari teks yang diberikan. Argumen yang diberikan dapat berupa string tunggal atau daftar string.

'input_ids': Ini adalah token-token yang dihasilkan dari tokenisasi teks input. Setiap token direpresentasikan sebagai ID numerik yang sesuai dalam kamus token yang digunakan oleh model. Di sini, 101 adalah ID untuk token [CLS], 7993 untuk token "Using", 170 untuk token "a", dan seterusnya, hingga 102 yang merupakan ID untuk token [SEP].

'token_type_ids': Ini adalah ID token jenis (token type IDs), yang menunjukkan jenis setiap token. Dalam model BERT, ketika menggunakan input single-sequence, semua token diberi label 0. Jadi, semua nilai dalam daftar ini adalah 0.

'attention_mask': Ini adalah masker perhatian (attention mask), yang menunjukkan bagian mana dari input yang sebenarnya berisi informasi yang relevan, dan bagian mana yang hanya berisi padding. Nilai 1 menunjukkan token-token yang sebenarnya, sedangkan nilai 0 menunjukkan token-token yang hanya padding dan seharusnya diabaikan oleh model. Dalam kasus ini, karena panjang sebenarnya dari teks adalah 7 token (tidak termasuk token [CLS] dan [SEP]), semua nilai dalam daftar ini adalah 1.

In [6]:
tokenizer.save_pretrained("directory_on_my_computer")

('directory_on_my_computer/tokenizer_config.json',
 'directory_on_my_computer/special_tokens_map.json',
 'directory_on_my_computer/vocab.txt',
 'directory_on_my_computer/added_tokens.json',
 'directory_on_my_computer/tokenizer.json')

In [7]:
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")

sequence = "Using a Transformer network is simple"
tokens = tokenizer.tokenize(sequence)

print(tokens)

['Using', 'a', 'Trans', '##former', 'network', 'is', 'simple']


In [8]:
ids = tokenizer.convert_tokens_to_ids(tokens)

print(ids)

[7993, 170, 13809, 23763, 2443, 1110, 3014]


In [9]:
decoded_string = tokenizer.decode([7993, 170, 11303, 1200, 2443, 1110, 3014])
print(decoded_string)

Using a transformer network is simple
